File: main/csharp/PrimitiveType.cs
namespace Adaptive.SimpleBinaryEncoding
{
    /// <summary>
    /// Primitive types from which all other types are composed.
    /// 
    /// <p/>
    /// <table>
    ///     <thead>
    ///         <tr>
    ///             <th>PrimitiveType Type</th>
    ///             <th>Description</th>
    ///             <th>Length (octets)</th>
    ///         </tr>
    ///     </thead>
    ///     <tbody>
    ///         <tr>
    ///             <td>char</td>
    ///             <td>Character</td>
    ///             <td>1</td>
    ///         </tr>
    ///         <tr>
    ///             <td>int8</td>
    ///             <td>Signed byte</td>
    ///             <td>1</td>
    ///         </tr>
    ///         <tr>
    ///             <td>uint8</td>
    ///             <td>Unsigned Byte / single byte character</td>
    ///             <td>1</td>
    ///         </tr>
    ///         <tr>
    ///             <td>int16</td>
    ///             <td>Signed integer</td>
    ///             <td>2</td>
    ///         </tr>
    ///         <tr>
    ///             <td>uint16</td>
    ///             <td>Unsigned integer</td>
    ///             <td>2</td>
    ///         </tr>
    ///         <tr>
    ///             <td>int32</td>
    ///             <td>Signed integer</td>
    ///             <td>4</td>
    ///         </tr>
    ///         <tr>
    ///             <td>uint32</td>
    ///             <td>Unsigned integer</td>
    ///             <td>4</td>
    ///         </tr>
    ///         <tr>
    ///             <td>int64</td>
    ///             <td>Signed integer</td>
    ///             <td>8</td>
    ///         </tr>
    ///         <tr>
    ///             <td>uint64</td>
    ///             <td>Unsigned integer</td>
    ///             <td>8</td>
    ///         </tr>
    ///         <tr>
    ///             <td>float</td>
    ///             <td>Single precision floating point</td>
    ///             <td>4</td>
    ///         </tr>
    ///         <tr>
    ///             <td>double</td>
    ///             <td>Double precision floating point</td>
    ///             <td>8</td>
    ///         </tr>
    ///     </tbody>
    /// </table>
    /// </summary>
    public class PrimitiveType
    {
        public static readonly PrimitiveType SbeChar = new PrimitiveType("char", 1, PrimitiveValue.MinValueChar, PrimitiveValue.MaxValueChar, PrimitiveValue.NullValueChar, SbePrimitiveType.Char);
        public static readonly PrimitiveType SbeInt8 = new PrimitiveType("int8", 1, PrimitiveValue.MinValueInt8, PrimitiveValue.MaxValueInt8, PrimitiveValue.NullValueInt8, SbePrimitiveType.Int8);
        public static readonly PrimitiveType SbeInt16 = new PrimitiveType("int16", 2, PrimitiveValue.MinValueInt16, PrimitiveValue.MaxValueInt16, PrimitiveValue.NullValueInt16, SbePrimitiveType.Int16);
        public static readonly PrimitiveType SbeInt32 = new PrimitiveType("int32", 4, PrimitiveValue.MinValueInt32, PrimitiveValue.MaxValueInt32, PrimitiveValue.NullValueInt32, SbePrimitiveType.Int32);
        public static readonly PrimitiveType SbeInt64 = new PrimitiveType("int64", 8, PrimitiveValue.MinValueInt64, PrimitiveValue.MaxValueInt64, PrimitiveValue.NullValueInt64, SbePrimitiveType.Int64);
        public static readonly PrimitiveType SbeUInt8 = new PrimitiveType("uint8", 1, PrimitiveValue.MinValueUint8, PrimitiveValue.MaxValueUint8, PrimitiveValue.NullValueUint8, SbePrimitiveType.UInt8);
        public static readonly PrimitiveType SbeUInt16 = new PrimitiveType("uint16", 2, PrimitiveValue.MinValueUint16, PrimitiveValue.MaxValueUint16, PrimitiveValue.NullValueUint16, SbePrimitiveType.UInt16);
        public static readonly PrimitiveType SbeUInt32 = new PrimitiveType("uint32", 4, PrimitiveValue.MinValueUint32, PrimitiveValue.MaxValueUint32, PrimitiveValue.NullValueUint32, SbePrimitiveType.UInt32);
        public static readonly PrimitiveType SbeUInt64 = new PrimitiveType("uint64", 8, PrimitiveValue.MinValueUint64, PrimitiveValue.MaxValueUint64, PrimitiveValue.NullValueUint64, SbePrimitiveType.UInt64);
        public static readonly PrimitiveType SbeFloat = new PrimitiveType("float", 4, PrimitiveValue.MinValueFloat, PrimitiveValue.MaxValueFloat, PrimitiveValue.NullValueFloat, SbePrimitiveType.Float);
        public static readonly PrimitiveType SbeDouble = new PrimitiveType("double", 8, PrimitiveValue.MinValueDouble, PrimitiveValue.MaxValueDouble, PrimitiveValue.NullValueDouble, SbePrimitiveType.Double);

        private readonly PrimitiveValue _maxVal;
        private readonly PrimitiveValue _minVal;
        private readonly string _name;
        private readonly PrimitiveValue _nullVal;
        private readonly int _size;
        private readonly SbePrimitiveType _sbePrimitiveType;

        internal PrimitiveType(string name, int size, long minVal, long maxVal, long nullVal, SbePrimitiveType sbePrimitiveType)
        {
            _name = name;
            _size = size;
            _sbePrimitiveType = sbePrimitiveType;
            _minVal = new PrimitiveValue(minVal, size);
            _maxVal = new PrimitiveValue(maxVal, size);
            _nullVal = new PrimitiveValue(nullVal, size);
        }

        internal PrimitiveType(string name, int size, double minVal, double maxVal, double nullVal, SbePrimitiveType sbePrimitiveType)
        {
            _name = name;
            _size = size;
            _sbePrimitiveType = sbePrimitiveType;
            _minVal = new PrimitiveValue(minVal, size);
            _maxVal = new PrimitiveValue(maxVal, size);
            _nullVal = new PrimitiveValue(nullVal, size);
        }

        /// <summary>
        ///     The name of the primitive type as a String.
        /// </summary>
        /// <value>the name as a String</value>
        public string PrimitiveName
        {
            get { return _name; }
        }

        /// <summary>
        ///     The size of the primitive type in octets.
        /// </summary>
        /// <value>size (in octets) of the primitive type</value>
        public int Size
        {
            get { return _size; }
        }

        /// <summary>
        ///     The minVal of the primitive type.
        /// </summary>
        /// <value>default minVal of the primitive type</value>
        public PrimitiveValue MinVal
        {
            get { return _minVal; }
        }

        /// <summary>
        ///     The maxVal of the primitive type.
        /// </summary>
        /// <value>default maxVal of the primitive type</value>
        public PrimitiveValue MaxVal
        {
            get { return _maxVal; }
        }

        /// <summary>
        ///     The nullVal of the primitive type.
        /// </summary>
        /// <value>default nullVal of the primitive type</value>
        public PrimitiveValue NullVal
        {
            get { return _nullVal; }
        }

        public SbePrimitiveType Type
        {
            get { return _sbePrimitiveType; }
        }
    }
}

File: sbe-tool/src/main/java/uk/co/real_logic/sbe/PrimitiveType.java
/*
 * Copyright 2013-2020 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe;

import static uk.co.real_logic.sbe.PrimitiveValue.*;

/**
 * Primitive types from which all other types are composed.
 */
public enum PrimitiveType
{
    CHAR("char", 1, MIN_VALUE_CHAR, MAX_VALUE_CHAR, NULL_VALUE_CHAR),
    INT8("int8", 1, MIN_VALUE_INT8, MAX_VALUE_INT8, NULL_VALUE_INT8),
    INT16("int16", 2, MIN_VALUE_INT16, MAX_VALUE_INT16, NULL_VALUE_INT16),
    INT32("int32", 4, MIN_VALUE_INT32, MAX_VALUE_INT32, NULL_VALUE_INT32),
    INT64("int64", 8, MIN_VALUE_INT64, MAX_VALUE_INT64, NULL_VALUE_INT64),
    UINT8("uint8", 1, MIN_VALUE_UINT8, MAX_VALUE_UINT8, NULL_VALUE_UINT8),
    UINT16("uint16", 2, MIN_VALUE_UINT16, MAX_VALUE_UINT16, NULL_VALUE_UINT16),
    UINT32("uint32", 4, MIN_VALUE_UINT32, MAX_VALUE_UINT32, NULL_VALUE_UINT32),
    UINT64("uint64", 8, MIN_VALUE_UINT64, MAX_VALUE_UINT64, NULL_VALUE_UINT64),
    FLOAT("float", 4, MIN_VALUE_FLOAT, MAX_VALUE_FLOAT, NULL_VALUE_FLOAT),
    DOUBLE("double", 8, MIN_VALUE_DOUBLE, MAX_VALUE_DOUBLE, NULL_VALUE_DOUBLE);

    private static final PrimitiveType[] VALUES = PrimitiveType.values();

    private final String name;
    private final int size;
    private final PrimitiveValue minValue;
    private final PrimitiveValue maxValue;
    private final PrimitiveValue nullValue;

    PrimitiveType(final String name, final int size, final long minValue, final long maxValue, final long nullValue)
    {
        this.name = name;
        this.size = size;
        this.minValue = new PrimitiveValue(minValue, size);
        this.maxValue = new PrimitiveValue(maxValue, size);
        this.nullValue = new PrimitiveValue(nullValue, size);
    }

    PrimitiveType(
        final String name, final int size, final double minValue, final double maxValue, final double nullValue)
    {
        this.name = name;
        this.size = size;
        this.minValue = new PrimitiveValue(minValue, size);
        this.maxValue = new PrimitiveValue(maxValue, size);
        this.nullValue = new PrimitiveValue(nullValue, size);
    }

    /**
     * The name of the primitive type as a String.
     *
     * @return the name as a String
     */
    public String primitiveName()
    {
        return name;
    }

    /**
     * The encodedLength of the primitive type in octets.
     *
     * @return encodedLength (in octets) of the primitive type
     */
    public int size()
    {
        return size;
    }

    /**
     * The minValue of the primitive type.
     *
     * @return default minValue of the primitive type
     */
    public PrimitiveValue minValue()
    {
        return minValue;
    }

    /**
     * The maxValue of the primitive type.
     *
     * @return default maxValue of the primitive type
     */
    public PrimitiveValue maxValue()
    {
        return maxValue;
    }

    /**
     * The nullValue of the primitive type.
     *
     * @return default nullValue of the primitive type
     */
    public PrimitiveValue nullValue()
    {
        return nullValue;
    }

    /**
     * Is the type an unsigned type like in C.
     *
     * @param type to be tested.
     * @return true if unsigned otherwise false.
     */
    public static boolean isUnsigned(final PrimitiveType type)
    {
        switch (type)
        {
            case UINT8:
            case UINT16:
            case UINT32:
            case UINT64:
                return true;

            default:
                return false;
        }
    }

    /**
     * Lookup PrimitiveType by String name and return Enum.
     *
     * @param name of primitiveType to get
     * @return the {@link PrimitiveType} matching the name
     * @throws IllegalArgumentException if name not found
     */
    public static PrimitiveType get(final String name)
    {
        for (final PrimitiveType primitiveType : VALUES)
        {
            if (primitiveType.name.equals(name))
            {
                return primitiveType;
            }
        }

        throw new IllegalArgumentException("No PrimitiveType for name: " + name);
    }
}
